content/lectures/lec02.md

```diff
@@ -7,3 +7,190 @@ draft: false
+
+## Demos
+
+### Docker
+
+As a short introduction to Docker, we're going to pull down a Docker image running Ubuntu and run that image in a container on our machine. To get started, first ensure that Docker Desktop is up and running on your machine. Then, we can pull the latest verision of Ubuntu, a Linux distribution which will run on our container (you can read more [here](https://hub.docker.com/_/ubuntu?tab=description&page=1&ordering=last_updated)):
+
+```
+$ docker pull ubuntu:latest
+latest: Pulling from library/ubuntu
+da7391352a9b: Pull complete 
+14428a6d4bcd: Pull complete 
+2c2d948710f2: Pull complete 
+Digest: sha256:c95a8e48bf88e9849f3e0f723d9f49fa12c5a00cfc6e60d2bc99d87555295e4c
+Status: Downloaded newer image for ubuntu:latest
+docker.io/library/ubuntu:latest
+```
+
+It's important to note here that when we tell Docker to pull the latest verison of Ubuntu, it actually does three seperate pulls from three seperate images. This is becauase Docker images are build inheriently from layers. Each pull is for a seperate layer of the image, and they are built on top of each other to produce the final product.
+
+When you think about it, this is a really smart way to create Docker images. Let's imagine a Docker container that runs Redis, the caching server that acts as a data store. To run Redis, we first need to have a Linux distribution installed, so we can add that as a lower layer of the Docker image which Redis then builds on top of by installing the specific Redis software.
+
+So, now that we have this Ubuntu Docker image installed, what can we do with it? Let's spin up the container and play around with it:
+
+```
+$ docker run -it --name ubuntu-demo ubuntu
+# This will open a terminal within the container running Ubuntu thanks to the "-it" flag
+# Now let's check to make sure we are actually running the latest version of Ubuntu
+root@260a8119abd6:/# cat /etc/os-release
+NAME="Ubuntu"
+VERSION="20.04.1 LTS (Focal Fossa)"
+ID=ubuntu
+ID_LIKE=debian
+PRETTY_NAME="Ubuntu 20.04.1 LTS"
+VERSION_ID="20.04"
+HOME_URL="https://www.ubuntu.com/"
+SUPPORT_URL="https://help.ubuntu.com/"
+BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
+PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
+VERSION_CODENAME=focal
+UBUNTU_CODENAME=focal
+```
+
+We can see that the Docker container is running a fresh install of Ubuntu. Now, we could do all manner of things with this Ubuntu container: install new software, expose ports, connect to a network, run a web server, etc. However, actually doing these things manually within the Ubuntu terminal is _not_ reproducible; we would have to perform those same operations every time we wanted to spin up a new container. Hence, a Dockerfile. When we write a Dockerfile it is essentially a roadmap for how the Docker image should be run, so that all of the installation and configuration of whatever toolchain we need on our container can be written out concisely once and only once.
```

Comment:
s/roadmap/instruction manual/

Reply:
Is that a directory on the ubuntu os with instructions?

Reply:
lol sry i meant replace the word "roadmap" with the word "instruction manual"